---
title: "Cumulative Regional Impact (CRI) Analysis Pipeline"
Author: Nickie Safarian
output: html_document
date: "2023-07-13"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# **R Markdown**
The result data generated in the previous step will be used here to further assess the distribution of Beta values in each region.
The null hypothesis is that there is no significant associations between genomic ranges/positions and distribution of beta values.

The effect size, defined as the contribution of a SNP to the genetic variance of the trait, is measured as beta coefficient (beta). 
The higher the absolute value of the beta coefficient, the stronger the effect. When density plotting for the avergae beta values per region,
we'll look for the skewness towards left (negative skewness) or right (right skewness), and coorelate that with the counts of snps per region.

## **1. Define some functions to calculate and plot quantiles**
This function will calculate quantiles in the following range: 
<0.00, 0.05, 0.10, 0.15, 0.20, 0.25, 0.30, 0.35, 0.40, 0.45, 0.50, 0.55, 0.60, 0.65, 0.70, 0.75, 0.80, 0.85, 0.90, 0.95, 1.00> .

```{r}
Quantil_Func <- function(x) {
  quantiles <- quantile(x, probs = seq(0, 1, 1/20)) 
  
  Iqr5_95 <- quantiles[20] - quantiles[2]
  Iqr25_75 <- quantiles[16] - quantiles[6]
  
  res <- c(quantiles, Iqr25_75, Iqr5_95)
  
  return(res)
}


HistPlot.Func <- function(x, Xlab){
  hist(x, probability = TRUE,
       col="lightgrey",
       xlab = Xlab,
       ylab = 'Frequency',
       main = 'Histogram')
  abline(v = mean(x), col='#2B0071', lwd = 3)
  lines(density(x), col = '#00416A', lwd = 3)
}

```

## **2. Calculate quantiles for the snp.counts**

```{r}
q_out <- Quantil_Func(result$snp.counts)
barplot(q_out[1:21]) 
```

The areas with extremely high or low counts of snps could be representative of HLA regions and pre-centromeric areas, respectively. 
We can exclude those intervals from further analysis.

## **3. Exclude ouliers (smallest 5% and biggest 5% counts)**

```{r}
filtr.dt <- result %>% 
  filter(snp.counts > quantile(snp.counts, 0.05) & snp.counts < quantile(snp.counts, 0.95)) 
```
Note how many snps are removed.

## **4. Calculate the counts zscore**

```{r}
filtr.dt  <- filtr.dt  %>% 
  mutate(ZCounts = as.numeric(scale(snp.counts)))
```

## **5. plot**

```{r}

HistPlot.Func(filtr.dt$snp.counts, Xlab="SNP RAW Counts \n After Removing Outliers")
HistPlot.Func(filtr.dt$ZCounts, Xlab="SNP Counts (scaled) \n After Filter")
```

## **6. Regression ANalysis: Does the snp counts coorelate with the average beta?**

```{r}
library(broom)
library(knitr)

# Linear modeling
lm_df = filtr.dt%>%
  # each step performed after this line is done with each cell type 
  group_by(CHR) %>%
  # fit cell type proportions according to this model  
  # using the broom package to tidy the results 
  do(tidy(lm(Avg.beta ~ scale(snp.counts), data = .))) %>%
  # ungroup the data
  ungroup() %>%
  # adjust for multiple comparisons using the Benjamini-Hochberg method
  mutate(padj = p.adjust(`p.value`, method = 'BH')) %>%
  # clean up variable names 
  mutate(term = recode(term, 
                       `(Intercept)` = "Intercept", 
                       `scale(snp.counts)` = "SNP_ZCounts"))



# print data frame for just age beta coefficients
kable(lm_df %>% filter(term == 'SNP_ZCounts'))

```

# Check for the direction of correlation for each chromosome.


## **7. Plot the coefficient of correlation per chromosome**

```{r}

library(RColorBrewer)

beta_plot = lm_df %>% 
  filter(! term =="Intercept") %>% 
  ggplot(aes(x = term, y = estimate, fill= term)) + 
  geom_hline(yintercept = 0) + 
  geom_bar(stat = "identity") + 
  geom_errorbar(aes(ymin = estimate - std.error, ymax = estimate + std.error)) + 
  ggtitle("Average Beta correlation with SNP counts per genomic range") +
  ylab('Std. Beta coeff.') + 
  xlab('Cell type proportions') + 
  theme(axis.text.x = element_text(size=10, face='bold', 
                                   angle = 90, vjust = 0.5, hjust=1))+
  scale_fill_brewer(palette="PuBuGn", direction = -1)
  #facet_wrap(~term, drop = T, scale = "free")
```

## **8. Simple scatter plot**

```{r}
library(ggpubr) # to add fitted regression equation inside
ggplot(data=filtr.dt, aes(x = snp.counts, y = Avg.beta)) + 
  geom_point() +
  stat_smooth(method = "lm", se=FALSE) +
  stat_regline_equation(label.x.npc = "center")+
  theme_bw()

```









